### PR TITLE
Support for composite indexes

### PR DESCRIPTION
**Pull request by jhchabran**
This pull-request adds support for composite indexes. It uses `document.Array` to wrap the multiple values into one for the index stores. The changes are not that large as the line changes count suggests, this is mainly due to having added a lot of table tests on the indexes. 

I have also rewritten the tests for the index iteration methods, which are now quite exhaustive. 

The [Benchmarks](https://gist.github.com/jhchabran/270e9baa00944b10a2ab43dcc297b369) are here and are quite good (0% change on simple indexes).

### CODE REVIEW

File: database/catalog.go
Comment (jhchabran) [inline database/catalog.go:313]: A quick change to make it work.

File: database/catalog.go
Comment (jhchabran) [inline database/catalog.go:482]: Same.

File: database/catalog_test.go
Comment (jhchabran) [inline database/catalog_test.go:116]: Same.

File: database/config.go
Comment (jhchabran) [inline database/config.go:617]: Same.

File: database/config.go
Comment (jhchabran) [inline database/config.go:622]: TODO @jhchabran fix this comment.

File: database/config.go
Comment (jhchabran) [inline database/config.go:691]: Same.

File: database/index.go
Comment (jhchabran) [inline database/index.go:23]: Here is the main work of the PR.

File: planner/optimizer.go
Comment (jhchabran) [inline planner/optimizer.go:360]: TODO @jhchabran, this change shouldn't be in this PR.

File: planner/optimizer.go
Comment (jhchabran) [inline planner/optimizer.go:544]: This makes it work with the existing code.

File: query/select_test.go
Comment (jhchabran) [inline query/select_test.go:35]: TODO @jhchabran drop this

File: stream/iterator.go
Comment (jhchabran) [inline stream/iterator.go:275]: That's the other part of the main work.

File: database/index.go
Comment (asdine) [inline database/index.go:189]: Actually, for n pivots, the n - 1 first must have a value I think.

```sql
CREATE INDEX idx_foo_a_b_c_d ON foo(a, b, c, d);

-- valid cases
WHERE a > 10 ;
WHERE a = 1 AND b > 10 ;
WHERE a = 1 AND b = 2 AND c > 10 ;
WHERE a = 1 AND b = 2 AND c = 3 AND d > 10 ;
```

Comment (jhchabran) [reply database/index.go:189]: These rules are defining what valid pivots are from the perspective of the index, not from the query planner. 

The no pivots at all case is there to support the `StreamScanOperator` case where we have a max but no min value, which means the indexed will be traversed until the max is reached: 

```
CREATE INDEX idx_foo_a_b_c_d ON foo(a, b, c, d);

WHERE a < 10 ;
```

In that case, the ranges are: `Min: nil, Max: [10]` which will lead to iterate without pivots.

File: planner/optimizer.go
Comment (jhchabran) [inline planner/optimizer.go:766]: Once merged, we should track this with an issue, at least not to forget it when making a release 😄

File: planner/optimizer.go
Comment (jhchabran) [inline planner/optimizer.go:402]: The new code is lengthy, but it felt easier to document and read in order to understand what's going.

File: database/index.go
Comment (asdine) [inline database/index.go:42]: 🕐 We should define a type for the zero value of ValueTypes

File: database/index.go
Comment (asdine) [inline database/index.go:71]:
```suggestion
// Set associates a list of values with a key. If Unique is set to false, it is
```

File: database/index.go
Comment (asdine) [inline database/index.go:192]: 🕐 We might want to fall in cases where there are no pivot but we want to read from the index anyway:

```sql
SELECT * FROM foo ORDER BY a [ASC|DESC]
```

If `a` is indexed, this will probably use the index to avoid having to sort the table (which is memory intensive). In this case, we won't have any range but will use the index to iterate over everything.

File: database/index.go
Comment (asdine) [inline database/index.go:190]: This shows that we might need a type for the pivot itself, and for the list of pivots as well, I think it will help understanding how pivots should work and separate pivot code from index code.

Something like this:
```go
type Pivot document.Value

func (p *Pivot) Validate() error

type Pivots []document.Value

func (p *Pivots) Validate() error
```

File: database/index.go
Comment (asdine) [inline database/index.go:242]: What does "empty" means in this case?

File: database/index.go
Comment (asdine) [inline database/index.go:323]: This comment is no longer true, leftover after rebase?

File: database/index.go
Comment (jhchabran) [reply database/index.go:42]: 🕐 What about: 
- having the zero value always invalid and never being equal to any type
- having an additional type `any` to use where we previously used the zero value (max int?)

**Review by asdine — APPROVED** [review]
I'm really excited about this, thanks @jhchabran !
I'm approving to unblock, feel free to merge once you feel it's ready

File: database/table.go
Comment (jhchabran) [reply database/table.go:222]: Fixing this, this is a leftover, I didn't properly log all my TODOs and forgot about this one.

File: query/reindex_test.go
Comment (jhchabran) [reply query/reindex_test.go:85]: Yes it does :)

File: planner/optimizer.go
Comment (jhchabran) [reply planner/optimizer.go:565]: We cannot, because the only other one is just for the primary key because it builds the candidate on the fly.

**Pull request merged**
